corehq/apps/users/audit/change_messages.py

```diff
@@ -1,91 +1,329 @@
+        else:
+            change_message = {"profile": {Change.SET: {}}}
+        return change_message
 
     @staticmethod
     def primary_location_removed():
-        return "Primary location: None"
+        return {"location": {Change.SET: {}}}
 
     @staticmethod
-    def commcare_user_primary_location_info(location_name):
-        return f"Primary location: {location_name}"
-
-    @staticmethod
-    def web_user_primary_location_info(location):
+    def primary_location_info(location):
         if location:
-            message = f"Primary location: {location.name}[{location.location_id}]"
+            change_message = {"location": {Change.SET: {"id": location.location_id, "name": location.name}}}
         else:
-            message = "Primary location: None"
-        return message
+            change_message = {"location": {Change.SET: {}}}
+        return change_message
 
     @staticmethod
-    def commcare_user_assigned_locations_info(location_names):
-        return f"Assigned locations: {location_names}"
-
-    @staticmethod
-    def web_user_assigned_locations_info(locations_info):
-        return f"Assigned locations: {locations_info}"
+    def assigned_locations_info(locations):
+        if locations:
+            change_message = {
+                "assigned_locations": {Change.SET: [
+                    {'id': location.location_id, 'name': location.name}
+                    for location in locations
+                ]}
+            }
+        else:
+            change_message = {"assigned_locations": {Change.SET: []}}
+        return change_message
 
     @staticmethod
-    def groups_info(groups_info):
-        return f"Groups: {groups_info}"
+    def groups_info(groups):
+        if groups:
+            change_message = {"groups": {Change.SET: [
+                {'id': group.get_id, 'name': group.name}
+                for group in groups
+            ]}}
+        else:
+            change_message = {"groups": {Change.SET: []}}
+        return change_message
 
     @staticmethod
     def added_as_web_user(domain):
-        return f"Added as web user to domain '{domain}'"
+        return {"domain": {Change.ADD: {"name": domain, "web_user": True}}}
 
     @staticmethod
     def invited_to_domain(domain):
-        return f"Invited to domain '{domain}'"
+        return {"domain_invitation": {Change.ADD: {'domain': domain}}}
 
     @staticmethod
     def invitation_revoked_for_domain(domain):
-        return f"Invitation revoked for domain '{domain}'"
+        return {"domain_invitation": {Change.REMOVE: {'domain': domain}}}
+
+
+class UserChangeMessageFormatterV1(object):
+    """
+    This class exposes just get_messages method which builds human readable
+    messages for structured messages created by UserChangeMessageV1.
+
+    It simply has a method _{change_slug}_messages for each change slug added by
+    UserChangeMessageV1
+    """
```

Comment:
> the methods are written defensively, and might return no messages without an explanation

I am thinking it should be okay if no message is returned. Are you saying it should fail if no message is returned?

The only place where i think it would be useful to be okay with empty messages is, when logs are versioned and say, if V2 adds more info but wants to build the message from V1 first, it should not break if there is no log for V1 to produce.
otherwise as of now, it would always create the message.

Reply:
> I am thinking it should be okay if no message is returned. Are you saying it should fail if no message is returned?

Yeah, either fail or display what we can, even if it's just the raw JSON or something.  Though I don't think that should ever need to happen, since we control what's in the system.

Reply:
It's currently not returning messages only if it's not meant to or if there is nothing to display based on what is set in the log.